File: code/python/ch17/practice/practice_04_find_median.py
"""
Practice 4: Find Median from Data Stream
============================================
Chapter 17: Heaps & Priority Queues — The VIP Line

PROBLEM
-------
Design a data structure that supports:
  - add_num(num): adds an integer from the stream
  - find_median(): returns the median of all added numbers

The function `solve` receives a list of numbers and returns a list of medians
after adding each number. For even counts, the median is the average of
the two middle values.

INPUT FORMAT
------------
A single line of space-separated integers.

OUTPUT FORMAT
-------------
A list of floats — the median after each insertion.

CONSTRAINTS
-----------
- 1 <= len(nums) <= 10^5
- -10^5 <= nums[i] <= 10^5

EXAMPLES
--------
Input:
  5 15 1 3
Output: [5.0, 10.0, 5.0, 4.0]

Input:
  2 3 4
Output: [2.0, 2.5, 3.0]

HINT
----
Use two heaps: a max-heap for the lower half and a min-heap for the
upper half. Balance them so they differ in size by at most 1.
The median is at the top of one or both heaps.

INSTRUCTIONS
------------
Replace the `pass` in the solve() function with your solution.
The main block at the bottom handles input/output — don't change it.
"""


def solve(nums: list[int]) -> list[float]:
    """Return a list of medians after adding each number."""
    pass  # TODO: Replace this with your solution


# ── Do not change anything below this line ──────────────────────────
if __name__ == "__main__":
    nums = list(map(int, input().strip().split()))
    print(solve(nums))

File: code/cpp/ch17/practice/practice_04_find_median.cpp
/*
 * Practice 4: Find Median from Data Stream
 * ===========================================
 * Chapter 17: Heaps & Priority Queues — The VIP Line
 *
 * PROBLEM:
 *   Given a stream of integers (as an array), return a list of medians
 *   after adding each number. Use the two-heap technique.
 *
 * EXAMPLES:
 *   solve({5,15,1,3})  -> {5.0, 10.0, 5.0, 4.0}
 *   solve({2,3,4})     -> {2.0, 2.5, 3.0}
 *
 * CONSTRAINTS:
 *   - 1 <= nums.size() <= 10^5
 *   - -10^5 <= nums[i] <= 10^5
 *
 * INSTRUCTIONS:
 *   Replace the body of solve() with your solution.
 */

#include <functional>
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

vector<double> solve(vector<int> nums) {
    // TODO: Replace this with your solution
    return {};
}

// -- Do not change anything below this line --------------------------
int main() {
    int n;
    cin >> n;
    vector<int> nums(n);
    for (int i = 0; i < n; i++) cin >> nums[i];
    vector<double> result = solve(nums);
    for (int i = 0; i < (int)result.size(); i++) {
        if (i) cout << " ";
        cout << result[i];
    }
    cout << endl;
    return 0;
}
